### PR TITLE
Feature/signup

### PR DESCRIPTION
**Pull request by enzofab91**
Trello reference:
https://trello.com/c/Y9W6nlnb

This PR generates sign up endpoint using device and device_auth_token gems. It also creates apiary documentation.

### CODE REVIEW

File: .circleci/database.yml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+test:
+ adapter: postgresql
+ encoding: unicode
+ pool: 5
+ username: postgres
+ password: postgres
+ host: localhost
+ port: 5432
+ database: test_db
```
Comment (sebastiancaraballo) [inline .circleci/database.yml:9]: Add a line here.

File: .rubocop.yml [hunk 2]
```diff
@@ -13,3 +13,25 @@ AllCops:
...
   Exclude:
     - "config/**/*"
+Layout/HashAlignment:
+  Exclude:
+    - "lib/**/*"
+Layout/LineLength:
+  Exclude:
+    - "config/**/*"
+    - db/migrate/20210413180512_add_devise_to_users.rb
+    - db/migrate/20210414141732_add_devise_token_auth_fields_users.rb
```
Comment (sebastiancaraballo) [inline .rubocop.yml:23]: Maybe exclude this cop from all migrations?

File: .rubocop.yml [hunk 3]
```diff
@@ -13,3 +13,25 @@ AllCops:
...
+Rails/RakeEnvironment:
+  Exclude:
+    - "lib/**/*"
+Metrics/BlockLength:
+  Exclude:
+    - config/**/*
+    - lib/**/*
+Style/SymbolArray:
+  Exclude:
+    - config/**/*
```
Comment (sebastiancaraballo) [inline .rubocop.yml:37]: Add a line here.

File: apiary.apib [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    + Body
+
+            {
+              "email":"user@example.com",
+              "password":"password",
+              "password_confirmation":"password_not_match",
+              "gender":"M"
+            }
+
++ Response 404 ()
```
Comment (sebastiancaraballo) [inline apiary.apib:23]:
```suggestion
+ Response 404
```

File: apiary.apib [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+                "password_confirmation": [
+                  "doesn't match Password"
+                ],
+                "full_messages": [
+                  "Password confirmation doesn't match Password"
+                ]
+              }
+            }
+
++ Request Ok
```
Comment (sebastiancaraballo) [inline apiary.apib:39]:
```suggestion
+ Request 200
```

File: apiary.apib [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    + Body
+
+            {
+              "email":"user@example.com",
+              "password":"password",
+              "password_confirmation":"password",
+              "gender":"M"
+            }
+
++ Response 200 ()
```
Comment (sebastiancaraballo) [inline apiary.apib:54]:
```suggestion
+ Response 200
```

File: apiary.apib [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+            {
+              "status": "success"",
+              "data": {
+                "email":"user@example.com",
+                "password":"password",
+                "password_confirmation":"password_not_match",
+                "gender":"M"
+              }
+            }
```
Comment (sebastiancaraballo) [inline apiary.apib:66]: Add a line here.

File: apiary.apib [hunk 8]
```diff
@@ -0,0 +1,66 @@
+FORMAT: 1A
+# Rails API Target
+
+# Group Users
+
+## Create User [api/v1/users/] [POST]
```
Comment (sebastiancaraballo) [inline apiary.apib:6]: Take a look https://apiblueprint.org/documentation/tutorial.html#resource

File: apiary.apib [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
++ Request Bad
+
+    + Headers
+
+            Content-Type: application/json
+
+    + Body
+
+            {
+              "email":"user@example.com",
```
Comment (sebastiancaraballo) [inline apiary.apib:17]: Look at this https://apiblueprint.org/documentation/advanced-tutorial.html#data-structures

File: spec/requests/api/v1/users/create_spec.rb [hunk 10]
```diff
@@ -0,0 +1,82 @@
+require 'rails_helper'
+
+describe 'User Api', type: :request do
```
Comment (sebastiancaraballo) [inline spec/requests/api/v1/users/create_spec.rb:3]:
```suggestion
describe 'POST api/v1/users', type: :request do
```

File: spec/requests/api/v1/users/create_spec.rb [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+          password_confirmation: password_confirmation,
+          gender: gender
+        }
+      }
+    end
+
+    it 'returns a successful response' do
+      post user_registration_path, params: params, as: :json
+
+      expect(response).to have_http_status(:success)
```
Comment (sebastiancaraballo) [inline spec/requests/api/v1/users/create_spec.rb:26]:
```suggestion
      expect(response).to be_successful
```

File: spec/requests/api/v1/users/create_spec.rb [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+          email: email,
+          password: password,
+          password_confirmation: password_confirmation,
+          gender: gender
+        }
+      }
+    end
+
+    it 'returns a successful response' do
+      post user_registration_path, params: params, as: :json
```
Comment (sebastiancaraballo) [inline spec/requests/api/v1/users/create_spec.rb:24]: Please, define this as a subject and then call it from the tests:
`subject { post user_registration_path, params: params, as: :json }`

File: spec/requests/api/v1/users/create_spec.rb [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+    it 'creates the user' do
+      expect {
+        post user_registration_path, params: params, as: :json
+      }.to change(User, :count).by(1)
+    end
+
+    it 'returns the user' do
+      post user_registration_path, params: params, as: :json
+
+      expect(response.body).to include_json(
```
Comment (sebastiancaraballo) [inline spec/requests/api/v1/users/create_spec.rb:38]: 👏

File: spec/requests/api/v1/users/create_spec.rb [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+      )
+    end
+
+    context 'when passwords dont match' do
+      let(:password_confirmation) { '12345' }
+
+      it 'does not return a successful response' do
+        post user_registration_path, params: params, as: :json
+
+        expect(response.status).to eq(failed_response)
```
Comment (sebastiancaraballo) [inline spec/requests/api/v1/users/create_spec.rb:57]:
```suggestion
        expect(response).to be_unprocessable_entity
```
Does that work?

**Review by sebastiancaraballo — COMMENTED** [review]
Great work! Left some comments. Please, review the code analysis check.

File: spec/requests/api/v1/users/create_spec.rb (hunk 14 above)
Comment (enzofab91) [reply spec/requests/api/v1/users/create_spec.rb:57]: That matcher makes test failed.

**Review by sebastiancaraballo — APPROVED** [review]
All good! Please ignore `.DS_Store` files.

**Pull request merged**